lib/Eisbach_Methods.thy

```diff
@@ -574,4 +575,156 @@ method if_concl for pattern::bool methods m =
+  let
+    val tac1 = method_evaluate m1 ctxt facts
+    val tac2 = method_evaluate m2 ctxt facts
+    (* Restricting tac1 is here for consistency with ";".
+       (all \<open>simp?\<close>; tac2) does not do what one might expect *)
+    val tac = ((K (restrict_goal 1 1 tac1)) FWD_ALL_NEW (fn i => restrict_goal i 1 tac2))
+  in
+    SIMPLE_METHOD (tac 1) facts
+  end)\<close>
+  \<open>forwards version of ";" - second method is applied to all goals produced by the first\<close>
+
+experiment begin
+
+lemma
+  assumes a: "\<And>x. P x"
+  assumes b: "\<And>x. Q x"
+  assumes c: "\<And>x. P x \<Longrightarrow> P x"
+  assumes goal_multiplicity2: "\<And>n. \<lbrakk> P (n+10);P (n+20) \<rbrakk> \<Longrightarrow> P n"
+  assumes goal_multiplicity5: "\<And>n. \<lbrakk> P (n+1);P (n+2);P (n+3);P (n+4);P (n+5) \<rbrakk> \<Longrightarrow> P n"
+  assumes goal_multiplicity2x: "\<And>n. \<lbrakk> P n;Q n \<rbrakk> \<Longrightarrow> P n"
+  shows "P 0"
+proof -
+  (* tests for generating a three-deep stack of rule production *)
+
+  (* generating using semicolon *)
+  have "P 0"
+  apply (rule goal_multiplicity2; (print_concl, rule goal_multiplicity5
+         (* invoked only on P goals, results in 20 subgoals *)
+         ; (print_concl, rule goal_multiplicity2x)
+         (* does not change number of goals *)
+         ; (rule c)?))
+    by (repeat 20 \<open>rule a b\<close>)
+
+  have "P 0"
+    apply (rule goal_multiplicity2; rule c)
+    apply (rule goal_multiplicity5; rule c)
+    apply (fails \<open>rule goal_multiplicity2x; rule c\<close>)
+    apply (succeeds \<open>rule goal_multiplicity2x; (rule c)?\<close>)
+    by (repeat 6 \<open>rule a b\<close>)
+
+  (* generating using fwd_all_new - using ALLGOALS_REV and goal restriction *)
+  have "P 0"
+    apply (
+      fwd_all_new \<open>rule goal_multiplicity2\<close>
+        \<open>print_concl, fwd_all_new \<open>rule goal_multiplicity5\<close>
+                                  \<open>print_concl, fwd_all_new \<open>rule goal_multiplicity2x\<close> \<open>(rule c)?\<close>\<close>\<close>)
+    by (repeat 20 \<open>rule a b\<close>)
+
+  have "P 0"
+    apply (fwd_all_new \<open>rule goal_multiplicity2\<close> \<open>rule c\<close>)
+    apply (fwd_all_new \<open>rule goal_multiplicity5\<close> \<open>rule c\<close>)
+    apply (fails \<open>fwd_all_new \<open>rule goal_multiplicity2x\<close> \<open>rule c\<close>\<close>)
+    apply (succeeds \<open>fwd_all_new \<open>rule goal_multiplicity2x\<close> \<open>(rule c)?\<close>\<close>)
+    by (repeat 6 \<open>rule a b\<close>)
+
+  (* check also that we can solve everything it spits out *)
+  have "P 0"
+    by (
+      fwd_all_new
+        \<open>fwd_all_new \<open>rule goal_multiplicity2\<close>
+          \<open>print_concl, fwd_all_new \<open>rule goal_multiplicity5\<close>
+          \<open>rule goal_multiplicity2x\<close>\<close>\<close>
+        \<open>rule a b\<close>)
+  oops
```

Comment:
Since we're in experiment, we could prove the lemma with `show "P 0" by (rule a)` to avoid the explicit oops.

Reply:
I did the oops on purpose, thinking there is no benefit to exporting more lemmas into the context. Still want me to change it?

Reply:
When you close `experiment` it won't be exported into the context. (I think that's the main reason Makarius added it, but maybe there is more).

Reply:
Result of investigation for `experiment`: it likely will be in the context, but inaccessible. A version that might not produce anything could be `notepad`. In any case, no big deal, it's also fine to just leave the `oops`.

Reply:
I changed it to be one big `notepad`, that way nothing gets exported, but I did have to add some `fix` to get it to work and not spit warnings.